ir/src/constraints/list_comprehension.rs

```diff
@@ -0,0 +1,328 @@
+        }
+        Expression::Exp(lhs, rhs) => {
+            let lhs = parse_lc_expr(lhs, lc, symbol_table, i)?;
+            let rhs = parse_lc_expr(rhs, lc, symbol_table, i)?;
+            Ok(Expression::Exp(Box::new(lhs), Box::new(rhs)))
+        }
+        Expression::ListFolding(lf_type) => {
+            parse_list_folding(lf_type, expression, symbol_table, i)
+        }
+        _ => Ok(expression.clone()),
+    }
+}
+
+/// Parses an identifier in a list comprehension expression.
+///
+/// # Errors
+/// - Returns an error if the iterable is an identifier and that identifier does not correspond to
+///   a vector.
+/// - Returns an error if the iterable is an identifier but is not of a type in set:
+///   { TraceColumns, IntegrityVariable, PublicInput, RandomValuesBinding }.
+/// - Returns an error if the iterable is a slice and that identifier does not correspond to
+///   a vector.
+/// - Returns an error if the iterable is an identifier but is not of a type in set:
+///   { TraceColumns, IntegrityVariable, PublicInput, RandomValuesBinding }.
+fn parse_elem(
+    ident: &Identifier,
+    expression: &Expression,
+    lc: &ListComprehension,
+    symbol_table: &SymbolTable,
+    i: usize,
+) -> Result<Expression, SemanticError> {
+    for (member, iterable) in lc.context() {
+        if ident == member {
+            match iterable {
+                Iterable::Identifier(ident) => {
+                    let ident_type = symbol_table.get_type(ident.name())?;
+                    match ident_type {
+                        IdentifierType::TraceColumns(trace_columns) => {
+                            validate_access(i, trace_columns.size())?;
+                            return Ok(Expression::NamedTraceAccess(NamedTraceAccess::new(
+                                ident.clone(),
+                                i,
+                                CURRENT_ROW,
+                            )));
+                        }
+                        IdentifierType::IntegrityVariable(var_type) => {
+                            match var_type.value() {
+                                VariableType::Vector(vector) => {
+                                    validate_access(i, vector.len())?;
+                                    return Ok(vector[i].clone());
+                                }
+                                // TODO: Handle matrix access
+                                _ => Err(SemanticError::InvalidListComprehension(
+                                    "Iterables should be vectors".to_string(),
+                                ))?,
+                            }
+                        }
+                        IdentifierType::PublicInput(size) => {
+                            validate_access(i, *size)?;
+                            return Ok(Expression::VectorAccess(VectorAccess::new(
+                                ident.clone(),
+                                i,
+                            )));
+                        }
```

Comment:
Public inputs should not be accessible in integrity constraints. Is this checked somewhere later?

Reply:
Thanks for catching this. It's not checked currently. I've added a check later on to throw an error if public inputs are accessed in integrity constraints.